RetroAchievements/RAWeb
# feat(game): sync tags to the taggables table

This PR begins the initial work of migrating tags (`~Hack~`, `~Homebrew~`, `~Prototype~`, etc) out of game titles and into structured data powered by `spatie/laravel-tags`.

There are two new behaviors included in this PR:
* A new command has been added to perform the initial sync of structured tags from `GameData`.
* Game title changes now double-write to structured tags.

**How to test this PR**
```bash
sail artisan ra:sync:legacy-game-tags 
```

After running the command, you should observe in your database that the `tags` and `taggables` tables are populated.

```sql
SELECT * FROM taggables WHERE taggable_type = "game" AND taggable_id = 1; -- returns 0 results
```

This should return 0 results.

Now, go change game ID 1's title to something like `"~Hack~ Sonic the Hedgehog"` and run the query again. You should get 1 result.

Later down the road, having tags as structured data will unlock several behaviors:
* Improved sorting and filtering by game tags.
* Internationalization of game tags.

This is also a precursor to migrating the beaten games leaderboard to React.

**NOTE:** A single game in your local snapshot probably has a tag of `~Multi~`. This game does not have a set. In the production database, I have removed this tag. It is not handled by the command or action.

Another thing to keep in mind is `spatie/laravel-tags` localizes tags, and there is no way to disable this behavior:
![Screenshot 2024-12-21 at 1 39 42 PM](https://github.com/user-attachments/assets/b34d5723-66bf-4447-8ce3-af828d92aec2)

We should probably only store "en" values in the database, and let the React UI do the presentation/localization work.
